Trippy/Utils/TSPUtils.m

```diff
@@ -0,0 +1,95 @@
+    }
+    NSArray *subarray = [elements subarrayWithRange:NSMakeRange(1, elements.count-1)];
+    NSMutableArray *results = [[NSMutableArray alloc] init];
+    for (NSArray *res in [self permutations:subarray]) {
+        for (int i = 0; i < elements.count; i+=1) {
+            NSArray *firstHalf = [res subarrayWithRange:NSMakeRange(0, i)];
+            NSArray *current = [elements subarrayWithRange:NSMakeRange(0, 1)];
+            NSArray *secondHalf = [res subarrayWithRange:NSMakeRange(i, res.count-i)];
+            NSArray *res = [[firstHalf arrayByAddingObjectsFromArray:current] arrayByAddingObjectsFromArray:secondHalf];
+            [results addObjectsFromArray:@[res]];
+        }
+    }
+    return results;
+}
+
++ (NSArray *)reorder:(NSArray *)elements order:(NSArray *)order {
+    NSMutableArray *newList = [[NSMutableArray alloc] init];
+    for (NSNumber *ix in order) {
+        [newList addObject:[elements objectAtIndex:[ix intValue]]];
+    }
+    return newList;
+}
+
++ (int)totalDistance:(NSArray *)order matrix:(NSDictionary *)matrix {
+    int sum = 0;
+    int firstIndex = [[order firstObject] intValue];
+    NSNumber *initialWeight = matrix[@"rows"][0][@"elements"][firstIndex+1][@"distance"][@"value"];
+    sum += [initialWeight intValue];
+    for (int i=1; i < order.count; i+=1) {
+        int current = [order[i] intValue];
+        int previous = [order[i-1] intValue];
+        NSArray *edges = matrix[@"rows"][previous+1][@"elements"];
+        NSNumber *weight = edges[current+1][@"distance"][@"value"];
+        sum += [weight intValue];
+    }
+    int lastIndex = [[order lastObject] intValue];
+    NSNumber *finalWeight = matrix[@"rows"][lastIndex+1][@"elements"][0][@"distance"][@"value"];
+    return sum + [finalWeight intValue];
+}
+
++ (int)totalDuration:(NSArray *)order matrix:(NSDictionary *)matrix {
+    int sum = 0;
+    int firstIndex = [[order firstObject] intValue];
+    NSNumber *initialWeight = matrix[@"rows"][0][@"elements"][firstIndex][@"duration"][@"value"];
+    sum += [initialWeight intValue];
+    for (int i=1; i < order.count; i+=1) {
+        int current = [order[i] intValue];
+        int previous = [order[i-1] intValue];
+        NSArray *edges = matrix[@"rows"][previous+1][@"elements"];
+        NSNumber *weight = edges[current+1][@"duration"][@"value"];
+        sum += [weight intValue];
+    }
+    int lastIndex = [[order lastObject] intValue];
+    NSNumber *finalWeight = matrix[@"rows"][lastIndex][@"elements"][0][@"duration"][@"value"];
+    return sum + [finalWeight intValue];
+}
+
+// brute force solution
++ (NSArray *)tspDistance:(NSDictionary *)matrix {
+    NSArray *locations = matrix[@"origin_addresses"];
+    int n = locations.count - 1;
+    
+    NSMutableArray *waypoints = [[NSMutableArray alloc] init];
+    for (int i = 0; i < n; i+=1) {
+        [waypoints addObject:[[NSNumber alloc] initWithInt:i]];
+    }
+    NSArray *potentialOrders = [self permutations:waypoints];
```

Comment:
It took me a bit to realize what this really is: it's essentially an array of arrays of indexes. Using `NSArray`s and `NSNumber`s feels a bit heavy handed; you could probably do a bit of old-school optimization by `malloc`ing an `int*` of the right size and filling it in, then making sure you index into it correctly 😄. Example:
```
int* list = malloc(sizeof(int)*6*3); // 6 total perms, 3 ints per perm
int row = 0;
list[(row*3)+0] = 0; list[(row*3)+1] = 1; list[(row*3)+2] = 2;
row++;
list[(row*3)+0] = 0; list[(row*3)+1] = 2; list[(row*3)+2] = 1;
row++;
list[(row*3)+0] = 1; list[(row*3)+1] = 0; list[(row*3)+2] = 2;
row++;
list[(row*3)+0] = 1; list[(row*3)+1] = 2; list[(row*3)+2] = 0;
row++;
list[(row*3)+0] = 2; list[(row*3)+1] = 0; list[(row*3)+2] = 1;
row++;
list[(row*3)+0] = 2; list[(row*3)+1] = 1; list[(row*3)+2] = 0;
```
(obviously you would stick with the recursive method you have to fill in the values, but just demonstrating the malloc and index into the array). Just don't forget to `free(list)` when you're done 😄.

Reply:
interesting...will also try to get around to optimizing this next week!